rasa_sdk/executor.py

```diff
@@ -176,54 +182,135 @@ def register_function(self, name: Text, f: Callable) -> None:
+
+        Args:
+            name: The module's name using Python's module path syntax.
+
+        Returns:
+            The loaded module.
+        """
+        module = importlib.import_module(name)
+
+        module_file = getattr(module, "__file__", None)
+        if module_file:
+            # If the module we're importing is a namespace package (a package
+            # without __init__.py), then there's nothing to watch for the
+            # package itself.
+            timestamp = os.path.getmtime(module_file)
+            self._modules[module_file] = (timestamp, module)
+
+        return module
+
     def register_package(self, package: Union[Text, types.ModuleType]) -> None:
-        from rasa_sdk.interfaces import Action
+        """Register all the `Action` subclasses contained in a Python module or
+        package.
 
+        Args:
+            package: Module or package name, or an already loaded Python module.
+        """
         try:
             self._import_submodules(package)
         except ImportError:
             logger.exception(f"Failed to register package '{package}'.")
             sys.exit(1)
 
+        self._register_all_actions()
+
+    def _register_all_actions(self) -> None:
+        """Scan for all user subclasses of `Action`, and register them.
+        """
+        import inspect
+
         actions = utils.all_subclasses(Action)
 
         for action in actions:
-            meta = action.__dict__.get("Meta", False)
-            abstract = getattr(meta, "abstract", False)
             if (
                 not action.__module__.startswith("rasa_core.")
                 and not action.__module__.startswith("rasa.")
                 and not action.__module__.startswith("rasa_sdk.")
                 and not action.__module__.startswith("rasa_core_sdk.")
-                and not abstract
+                and not inspect.isabstract(action)
             ):
                 self.register_action(action)
 
+    def reload(self) -> None:
+        """Reload all Python modules that have been loaded in the past.
+
+        To check if a module should be reloaded, the file's last timestamp is
+        checked against the current one. If the current last-modified timestamp
+        is larger than the previous one, the module is re-imported using
+        `importlib.reload`.
+
+        If one or more modules are reloaded during this process, the entire
+        `Action` class hierarchy is scanned again to see what new classes can
+        be registered.
+        """
+        to_reload = {}
+
+        for path, (timestamp, module) in self._modules.items():
+            try:
+                new_timestamp = os.path.getmtime(path)
+            except OSError:
+                # Ignore missing files
+                continue
+
+            if new_timestamp > timestamp:
+                to_reload[path] = (new_timestamp, module)
+
+        module_reloaded = False
```

Comment:
?
```suggestion
        all_modules_reloaded = False
```

Reply:
`any_modules_reloaded` is better I think, since we only care if at least one was reloaded